### PR TITLE
POA VACOLS Integration

### PR DESCRIPTION
**Pull request by sharonwarner**
Connects #2098 

This PR updates VACOLS POA information if the user says VACOLS isn't correct. 

- [ ] Confirmed the flow works correctly in test mode
- [ ] Confirmed VACOLS is updated in UAT in all four cases:
            - Address has to be updated
            - User first/last/middle name has to be updated
            - User first/last name has to be updated
            - Just the service organization has to be updated

### CODE REVIEW

File: app/models/vacols/representative.rb [hunk 1]
```diff
@@ -1,4 +1,72 @@
 class VACOLS::Representative < VACOLS::Record
   self.table_name = "vacols.rep"
   self.primary_key = "repkey"
+
+  class InvalidRepTypeError < StandardError; end
+
+  def self.update_vacols_rep_type!(case_record:, rep_type:)
+    return unless rep_type
+
+    # fail(InvalidRepTypeError) unless VALID_UPDATE_REP_TYPES.include?(rep_type)
```
Comment (amprokop) [inline app/models/vacols/representative.rb]: instead, how about

`VACOLS::CASE::REPRESENTATIVES.include?(rep_type)`

File: app/controllers/certifications_controller.rb [hunk 2]
```diff
@@ -58,6 +58,13 @@ def certify_v2
       certifying_official_title: certification.certifying_official_title
     )
     form8.save_pdf!
+    unless @certification.poa_correct_in_vacols
+      Appeal.repository.update_vacols_rep_info!(
```
Comment (amprokop) [inline app/controllers/certifications_controller.rb]: so this would be `appeal.power_of_attorney.update_vacols_rep_info!`

**Comment by amprokop** [comment]
@sharonwarner and i talked about some changes in Slack:

1. move the appeal repository code into power_of_attorney_repository
2. move the vacols interaction code into separate methods in that repository
3. stub those methods out in lib/fakes/power_of_attorney_repository
4. unit test the rest where possible

File: app/models/vacols/representative.rb [hunk 3]
```diff
@@ -1,4 +1,72 @@
...
+        conn.execute(<<-SQL)
+          UPDATE BRIEFF
+          SET BFSO = #{rep_type}
+          WHERE BFKEY = #{case_id}
+        SQL
+      end
+    end
+  end
+
+  def self.update_vacols_rep_name!(case_record:, first_name:, middle_initial:, last_name:)
```
Comment (amprokop) [inline app/models/vacols/representative.rb]: So it looks like we're not handling creation of a REP row, right?

File: app/services/appeal_repository.rb [hunk 4]
```diff
@@ -203,6 +203,61 @@ def self.update_vacols_after_dispatch!(appeal:, vacols_note: nil)
...
+  def self.get_vacols_reptype_code(short_name:)
+    VACOLS::Case::REPRESENTATIVES.each do |representative|
+      return representative[0] if representative[1][:short] == short_name
+    end
+    nil
+  end
+
+  def self.first_last_name?(representative_name:)
+    representative_name.split(" ").length == 2 &&
+      !representative_name.include?("-") &&
```
Comment (amprokop) [inline app/services/appeal_repository.rb]: why this? last names are sometimes hyphenated no?

File: app/services/appeal_repository.rb [hunk 5]
```diff
@@ -203,6 +203,61 @@ def self.update_vacols_after_dispatch!(appeal:, vacols_note: nil)
...
+      )
+    elsif Appeal.repository.first_middle_last_name?(representative_name: representative_name)
+      VACOLS::Representative.update_vacols_rep_name!(
+        case_record: appeal.case_record,
+        first_name: representative_name.split(" ")[0],
+        middle_initial: representative_name.split(" ")[1],
+        last_name: representative_name.split(" ")[2]
+      )
+    else
+      VACOLS::Representative.update_vacols_rep_address_one!(
```
Comment (amprokop) [inline app/services/appeal_repository.rb]: note that this is going to get more complicated with the ticket to actually update the address.

File: app/models/vacols/representative.rb (hunk 3 above)
Comment (sharonwarner) [inline app/models/vacols/representative.rb]: Yes, sorry, I should have talked to you about this. If a REP row doesn't exist, we're currently getting an error on the confirm_case_details page. Do we have an issue out to fix that? Maybe we could combine that issue with this?

File: app/services/appeal_repository.rb (hunk 4 above)
Comment (sharonwarner) [inline app/services/appeal_repository.rb]: Good point. Should I update that in the acceptance criteria as well?

File: app/controllers/certifications_controller.rb [hunk 6]
```diff
@@ -58,6 +58,13 @@ def certify_v2
       certifying_official_title: certification.certifying_official_title
     )
     form8.save_pdf!
+    unless @certification.poa_correct_in_vacols
+      PowerOfAttorney.repository.update_vacols_rep_info!(
```
Comment (amprokop) [inline app/controllers/certifications_controller.rb]: I think the certification should only know about the appeal, which knows about power of attorney model, which knows about the repository. The Power Of Attorney model should handle all interaction with other systems, we should preserve that boundary lest everything be tied to everything else.

so `certification.appeal.power_of_attorney.some_method`

File: app/models/vacols/representative.rb [hunk 7]
```diff
@@ -1,4 +1,72 @@
 class VACOLS::Representative < VACOLS::Record
   self.table_name = "vacols.rep"
   self.primary_key = "repkey"
+
+  class InvalidRepTypeError < StandardError; end
+
+  def self.update_vacols_rep_type!(case_record:, rep_type:)
+    return unless rep_type
```
Comment (amprokop) [inline app/models/vacols/representative.rb]: maybe we should fail here too? do you have some situation in mind where this is likely to be called with no rep_type?

File: app/models/vacols/representative.rb (hunk 3 above)
Comment (amprokop) [inline app/models/vacols/representative.rb]: @sharonwarner — think that error is handled by https://github.com/department-of-veterans-affairs/caseflow/pull/2235?

Comment (amprokop) [inline app/models/vacols/representative.rb]: let's make a separate PR to add the creation of a REP row.

Comment (amprokop) [inline app/models/vacols/representative.rb]: (so this unblocks @aroltsch on https://github.com/department-of-veterans-affairs/caseflow/pull/2235 faster)

File: app/services/power_of_attorney_repository.rb [hunk 8]
```diff
@@ -33,4 +33,81 @@ def self.set_vacols_values(poa:, case_record:)
...
+  def self.update_vacols_rep_address_one!(case_record:, address_one:)
+    VACOLS::Representative.update_vacols_rep_address_one!(
+      case_record: case_record,
+      address_one: address_one
+    )
+  end
+  # :nocov:
+
+  def self.update_vacols_rep_table(appeal:, representative_name:)
+    if first_last_name?(representative_name: representative_name)
```
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: do we need named parameters if there's just one parameter? seems like you use em a lot. i usually think for 3+ parameters they're useful but just more verbose for fewer

File: app/models/vacols/representative.rb [hunk 9]
```diff
@@ -1,4 +1,72 @@
...
+          SET REPFIRST = #{first_name},
+              REPMI = #{middle_initial},
+              REPLAST = #{last_name}
+          WHERE REPKEY = #{case_id}
+        SQL
+      end
+    end
+  end
+
+  def self.update_vacols_rep_address_one!(case_record:, address_one:)
```
Comment (amprokop) [inline app/models/vacols/representative.rb]: Not a huge deal, but it's usually preferable to pass as little as possible into a method.

So if we only need bfkey, maybe we only pass that instead of case_record. Up to you whether you want to change that or not. It's easier to test and clearer about the internals of the method. If you think we might later need more information from case_record, it could be good to pass the whole thing in.

File: lib/fakes/power_of_attorney_repository.rb [hunk 10]
```diff
@@ -14,4 +14,16 @@ def self.representative
   def self.poa_query(_poa)
     FakePoaRecord
   end
+
+  def self.update_vacols_rep_type!(_case_record, _vacols_rep_type)
+    nil
```
Comment (amprokop) [inline lib/fakes/power_of_attorney_repository.rb]: could be good to manually test or unit test what happens if there's an error on one of these updates.

File: spec/services/power_of_attorney_repository_spec.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
+describe PowerOfAttorneyRepository do
+  context ".get_vacols_reptype_code" do
+    subject { PowerOfAttorney.repository.get_vacols_reptype_code(short_name: short_name) }
+
+    context "returns a the VACOLS code when it exists" do
```
Comment (amprokop) [inline spec/services/power_of_attorney_repository_spec.rb]: typo `a the`

File: app/services/power_of_attorney_repository.rb [hunk 12]
```diff
@@ -33,4 +33,81 @@ def self.set_vacols_values(poa:, case_record:)
...
+
+  def self.get_vacols_reptype_code(short_name:)
+    VACOLS::Case::REPRESENTATIVES.each do |representative|
+      return representative[0] if representative[1][:short] == short_name
+    end
+    nil
+  end
+
+  def self.first_last_name?(representative_name:)
+    representative_name.split(" ").length == 2 && !representative_name.include?("&")
```
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: maybe `trim` in case of leading/trailing whitespace too

File: app/services/power_of_attorney_repository.rb [hunk 13]
```diff
@@ -33,4 +33,81 @@ def self.set_vacols_values(poa:, case_record:)
...
+  end
+
+  def self.first_middle_last_name?(representative_name:)
+    representative_name.split(" ").length == 3 && representative_name.split(" ")[1].tr(".", "").length == 1
```
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: maybe trim in case of leading/trailing whitespace too

File: app/services/power_of_attorney_repository.rb [hunk 14]
```diff
@@ -33,4 +33,81 @@ def self.set_vacols_values(poa:, case_record:)
...
+
+  def self.update_vacols_rep_address_one!(case_record:, address_one:)
+    VACOLS::Representative.update_vacols_rep_address_one!(
+      case_record: case_record,
+      address_one: address_one
+    )
+  end
+  # :nocov:
+
+  def self.update_vacols_rep_table(appeal:, representative_name:)
```
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: can we add a TODO here to revisit this logic? we may want to later stuff everything in the last name or somesuch

File: app/services/power_of_attorney_repository.rb [hunk 15]
```diff
@@ -33,4 +33,81 @@ def self.set_vacols_values(poa:, case_record:)
...
+  end
+
+  def self.update_vacols_rep_info!(appeal:, representative_type:, representative_name:)
+    if representative_type == "Service Organization"
+      # We set the rep type to the service organization name, unless we don't have a record
+      # of it. Then we set it to 'other'.
+      vacols_rep_type = get_vacols_reptype_code(short_name: representative_name) ||
+                        get_vacols_reptype_code(short_name: "Other")
+    else
+      vacols_rep_type = get_vacols_reptype_code(short_name: representative_type)
```
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: thinking out loud:

our representative types are:
```
Attorney
Agent
One Time Rep
Service Organization
None
```

so this should work

File: app/services/power_of_attorney_repository.rb [hunk 16]
```diff
@@ -33,4 +33,81 @@ def self.set_vacols_values(poa:, case_record:)
...
+    end
+    update_vacols_rep_type!(case_record: appeal.case_record, vacols_rep_type: vacols_rep_type)
+
+    if %w(T U O).include? vacols_rep_type
```
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: Should One time Rep be included in here too?

Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: I added `rep_name_in_rep_table` to VACOLS::CASE::REPRESENTATIVES so we could keep this info in one place, can we use that?

File: app/services/power_of_attorney_repository.rb (hunk 14 above)
Comment (amprokop) [inline app/services/power_of_attorney_repository.rb]: is there a reason why everything else has a `!` but this one doesn't?

File: app/services/power_of_attorney_repository.rb (hunk 8 above)
Comment (sharonwarner) [inline app/services/power_of_attorney_repository.rb]: Yeah, sorry... I'm obsessed with them. Lol. I'll take out the ones for just one variable.

File: app/controllers/certifications_controller.rb [hunk 17]
```diff
@@ -58,7 +58,14 @@ def certify_v2
       certifying_official_title: certification.certifying_official_title
     )
     form8.save_pdf!
-    @certification.complete!(current_user.id)
+    unless certification.poa_correct_in_vacols
```
Comment (amprokop) [inline app/controllers/certifications_controller.rb]: ahhh. i'm really sorry to be a nit picking monster, but i think this could maybe go in `certification.complete!`???

Comment (amprokop) [inline app/controllers/certifications_controller.rb]: it seems like it should live in the model rather than the controller (too much logic 4 tha controller)

File: app/models/power_of_attorney.rb [hunk 18]
```diff
@@ -60,6 +60,23 @@ def overwrite_vacols_with_bgs_value
     # case_record.bfso
   end
 
+  def update_vacols_rep_info!(appeal:, representative_type:, representative_name:)
+    repo = PowerOfAttorney.repository
+    vacols_rep_type = if representative_type == "Service Organization"
+                        # We set the rep type to the service organization name, unless we don't have a record
+                        # of it. Then we set it to 'other'.
+                        repo.get_vacols_reptype_code(representative_name) ||
```
Comment (amprokop) [inline app/models/power_of_attorney.rb]: nit: rep and type are still two words right? so maybe this should be`get_vacols_rep_type_code` or just `get_vacols_rep_code` since the values in that dropdown don't really exclusively correspond to name or type
